event-loop.go

```diff
@@ -0,0 +1,120 @@
+package flutter
+
+import (
+	"container/heap"
+	"fmt"
+	"math"
+	"runtime"
+	"time"
+
+	"github.com/go-flutter-desktop/go-flutter/embedder"
+	"github.com/go-flutter-desktop/go-flutter/internal/priorityqueue"
+)
+
+// EventLoop is a event loop for the main thread that allows for delayed task
+// execution.()
+type EventLoop struct {
+	// store the task (event) by their priorities
+	priorityqueue *priorityqueue.PriorityQueue
+	// called when a task has been received, used to Wakeup the rendering event loop
+	postEmptyEvent func()
+
+	onExpiredTask func(*embedder.FlutterTask) embedder.Result
+
+	// timeout for non-Rendering events that needs to be processed in a polling manner
+	maxWaitDuration time.Duration
+}
+
+func newEventLoop(postEmptyEvent func(), onExpiredTask func(*embedder.FlutterTask) embedder.Result) *EventLoop {
+	pq := priorityqueue.NewPriorityQueue()
+	heap.Init(pq)
+	return &EventLoop{
+		priorityqueue:  pq,
+		postEmptyEvent: postEmptyEvent,
+		onExpiredTask:  onExpiredTask,
+
+		// 25 Millisecond is arbitrary value, not too high (adds too much delay to
+		// platform messages) and not too low (heavy CPU consumption).
+		// This value isn't related to FPS, as rendering events are process in a
+		// waiting manner.
+		maxWaitDuration: time.Duration(25) * time.Millisecond,
+	}
+}
+
+// RunOnCurrentThread returns if the current thread is the thread used
+// by the GLFW event loop.
+func (t *EventLoop) RunOnCurrentThread() bool {
```

Comment:
This is given to `engine.TaskRunnerRunOnCurrentThread`, but what if it is called from C-land on a different Thread??

Sure, Application.Run() calls `runtime.LockOSThread`, but that doesn't guarantee that the Engine doesn't start it's own threads and calls this function from then, right?

Reply:
Agreed, the flutter GLFW shell uses `std::this_thread::get_id() == main_thread_id_;`
There aren't any api on golang to retrieve the current thread id, (well they are but if you use them [“If you use this package, you will go straight to hell.“](https://godoc.org/github.com/davecheney/junk/id))

Giving the facts that:
 - the engine can do the `std::this_thread::get_id() == main_thread_id_;` internally to check if he is running on his correct thread.   
- we only use one thread for the Platform Task Runner and UI Task Runner
`engine.TaskRunnerRunOnCurrentThread` [flutter/wiki](https://github.com/flutter/flutter/wiki/The-Engine-architecture#threading)
- IIRC, calls from the engine to the embedder are done in the same thread.

From my understanding of the golang language, there is no way of getting the thread_id, that why I made the choice to always return true.

If you have any proposition, please let me know!

Reply:
more info on https://github.com/flutter/engine/pull/8273#discussion_r269809503

Reply:
more info: https://github.com/flutter/engine/pull/8273#discussion_r269809503

Reply:
Resolved using C call